packages/appcache-polyfill-window/index.ts

```diff
@@ -0,0 +1,222 @@
+  if (manifestAttribute && 'serviceWorker' in navigator) {
+    const manifestUrl = (new URL(manifestAttribute, location.href)).href;
+
+    try {
+      const hash = await checkManifestVersion(manifestUrl);
+      // hash might be null if the manifest was removed.
+      if (hash) {
+        await updateManifestAssociationForCurrentPage(manifestUrl, hash);
+      } else {
+        await removeManifestAssociations(manifestUrl);
+      }
+    } catch (error) {
+      console.error('Unable to update App Cache associations:', error);
+    }
+  }
+}
+
+async function addToCache(hash: string, urls: Array<string>) {
+  const cache = await caches.open(hash);
+
+  await Promise.all(urls.map(async (url) => {
+    // See Item 18.3 of https://html.spec.whatwg.org/multipage/browsers.html#downloading-or-updating-an-application-cache
+    const init: RequestInit = {
+      credentials: 'include',
+      headers: [['X-Use-Fetch', 'true']],
+      redirect: 'manual',
+    };
+    const request = new Request(url, init);
+
+    try {
+      const response = await fetch(request);
+
+      const cacheControl = response.headers.get('Cache-Control');
+      if (cacheControl && cacheControl.includes('no-store')) {
+        // Bail early if we're told not to cache this response.
+        return;
+      }
+
+      if (response.status === 200) {
+        await cache.put(url, response);
+        return;
+      }
+
+      // See Item 18.5 of https://html.spec.whatwg.org/multipage/browsers.html#downloading-or-updating-an-application-cache
+      if (response.status !== 404 && response.status !== 410) {
+        // Assuming this isn't a 200, 404 or 410, we want the catch to
+        // trigger, which will cause any previously cached Response for this
+        // URL to be copied over to this new cache.
+        throw new Error(`Bad status: ${response.status}`);
+      }
+    } catch (e) {
+      // We're here if one of the following happens:
+      // - The fetch() rejected due to a NetworkError.
+      // - The HTTP status code from the fetch() was something other than
+      //   200, 404, and 410 AND the response isn't Cache-Control: no-store
+      const response = await caches.match(url);
+      if (response) {
+        await cache.put(url, response);
+      }
+    }
+  }));
+}
+
+async function checkManifestVersion(manifestUrl: string) {
+  // See Item 4 of https://html.spec.whatwg.org/multipage/browsers.html#downloading-or-updating-an-application-cache
+  const init: RequestInit = {
+    credentials: 'include',
+    headers: [['X-Use-Fetch', 'true']],
+  };
+  const manifestRequest = new Request(manifestUrl, init);
+
+  let manifestResponse = await fetch(manifestRequest);
+  // See Item 6 of https://html.spec.whatwg.org/multipage/offline.html#downloading-or-updating-an-application-cache
+  if (manifestResponse.status === 404 || manifestResponse.status === 410) {
+    // TODO: Double check that this is a reasonable approach.
```

Comment:
I'm really not sure about this bit. 100% fidelity with the App Cache spec may not be necessary as someone removing their App Cache manifest sounds while they're also using the polyfill sounds like they're trying to migrate off of App Cache.